File: clients/intellij-community/src/main/java/dev/cleanroommc/workbench/intellij/community/SourceNavigationClient.java
package dev.cleanroommc.workbench.intellij.community;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import java.net.URI;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.util.HexFormat;
import java.util.List;

/** Source-only transport and exact byte/UTF-16 location checks; no domain parser. */
final class SourceNavigationClient {
    private SourceNavigationClient() { }

    static JsonObject invoke(CoreLaunch launch, Path workspace, String session, List<String> action) throws Exception {
        if (!launch.host().equals("native")) throw new IllegalArgumentException("Source navigation requires a native Linux Workbench host.");
        if (session.isBlank() || session.startsWith("-") || session.length() > 256 || session.matches(".*\\s.*")) {
            throw new IllegalArgumentException("Enter one exact Work Session ID.");
        }
        var arguments = new java.util.ArrayList<>(List.of("context", "run", session, "--", "source"));
        arguments.addAll(action);
        String raw = CommandProcess.capture(launch, arguments, 16 * 1024 * 1024, 120, workspace.toString());
        JsonObject envelope = JsonParser.parseString(raw).getAsJsonObject();
        JsonObject result = envelope.getAsJsonObject("result");
        Path root = workspace.toRealPath();
        if (!envelope.get("format").getAsString().equals("workbench-developer-action-v1")
                || envelope.get("exit_code").getAsInt() != 0
                || !Path.of(URI.create(envelope.getAsJsonObject("context").getAsJsonObject("selection").get("pack_uri").getAsString())).equals(root)
                || !Path.of(URI.create(result.getAsJsonObject("context").getAsJsonObject("binding").getAsJsonObject("source_observation").get("root_uri").getAsString())).equals(root)
                || !result.getAsJsonObject("context").getAsJsonObject("authority").get("runtime_authority").getAsString().equals("none")) {
            throw new IllegalArgumentException("Source navigation changed the workspace or source-only authority.");
        }
        return result;
    }

    static Target verify(Path workspace, JsonObject location) throws Exception {
        String relative = location.get("path").getAsString();
        if (!location.get("coordinate_system").getAsString().equals("one-based-utf16")
                || !location.get("interval").getAsString().equals("half-open")
                || relative.isEmpty() || relative.contains("\\") || relative.contains(":") || relative.indexOf('\0') >= 0) {
            throw new IllegalArgumentException("Invalid source location.");
        }
        Path target = workspace.toRealPath();
        for (String part : relative.split("/", -1)) {
            if (part.isEmpty() || part.equals(".") || part.equals("..")) throw new IllegalArgumentException("Unsafe source path.");
            target = target.resolve(part);
            if (Files.isSymbolicLink(target)) throw new IllegalArgumentException("Source path traverses a symbolic link.");
        }
        if (!Files.isRegularFile(target) || Files.size(target) > 64 * 1024 * 1024) throw new IllegalArgumentException("Source file exceeds its bound.");
        byte[] raw = Files.readAllBytes(target);
        if (!HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(raw)).equals(location.get("sha256").getAsString())) {
            throw new IllegalArgumentException("Source selection is stale; search again.");
        }
        int start = location.get("byte_start").getAsBigDecimal().intValueExact();
        int end = location.get("byte_end").getAsBigDecimal().intValueExact();
        if (start < 0 || start > end || end > raw.length) throw new IllegalArgumentException("Invalid source interval.");
        for (String name : List.of("start", "end")) {
            String prefix = StandardCharsets.UTF_8.newDecoder().decode(ByteBuffer.wrap(raw, 0, name.equals("start") ? start : end)).toString();
            JsonObject point = location.getAsJsonObject(name);
            if (point.get("line").getAsBigDecimal().intValueExact() != prefix.chars().filter(c -> c == '\n').count() + 1
                    || point.get("column").getAsBigDecimal().intValueExact() != prefix.length() - prefix.lastIndexOf('\n')) {
                throw new IllegalArgumentException("Editor coordinates differ from exact source bytes.");
            }
        }
        String text = StandardCharsets.UTF_8.newDecoder().decode(ByteBuffer.wrap(raw)).toString().replace("\r\n", "\n");
        return new Target(target, text, location.deepCopy());
    }

    record Target(Path path, String text, JsonObject location) { }
}

File: clients/vscode/sourceNavigationClient.js
"use strict";

const fs = require("node:fs");
const path = require("node:path");
const { createHash } = require("node:crypto");
const { pathToFileURL } = require("node:url");
const { invokeCoreJson } = require("./coreCommandClient");
const { resolveCoreLaunch } = require("./coreLaunch");

async function invokeSourceAction(executable, session, args, options = {}) {
  if (typeof session !== "string" || !session || session.startsWith("-") || session.length > 256 || /[\s\0]/.test(session)) {
    throw new Error("Enter one exact Work Session ID from workbench context select.");
  }
  const launch = options.launch || resolveCoreLaunch(executable, options);
  if (launch.host !== "native") throw new Error("Source navigation currently requires a native Linux Workbench host.");
  const value = await invokeCoreJson(executable, ["context", "run", session, "--", "source", ...args], {
    ...options, launch, maximumOutput: 16 * 1024 * 1024, timeoutMs: 120000, label: "Source navigation",
  });
  const root = pathToFileURL(fs.realpathSync(options.cwd)).href;
  if (value.format !== "workbench-developer-action-v1" || value.exit_code !== 0
      || value.context?.selection?.pack_uri !== root
      || value.result?.context?.binding?.source_observation?.root_uri !== root
      || value.result.context.authority?.runtime_authority !== "none") {
    throw new Error("Source navigation changed the selected workspace or source-only authority.");
  }
  return value.result;
}

function verifiedSourceTarget(workspace, location) {
  if (!location || location.coordinate_system !== "one-based-utf16" || location.interval !== "half-open"
      || typeof location.path !== "string" || !location.path || /[\\:\0]/.test(location.path)
      || location.path.split("/").some((part) => !part || part === "." || part === "..")
      || !/^[0-9a-f]{64}$/.test(location.sha256)) throw new Error("Invalid source location.");
  let target = fs.realpathSync(workspace);
  for (const part of location.path.split("/")) {
    target = path.join(target, part);
    if (fs.lstatSync(target).isSymbolicLink()) throw new Error("Source location traverses a symbolic link.");
  }
  const metadata = fs.statSync(target);
  if (!metadata.isFile() || metadata.size > 64 * 1024 * 1024) throw new Error("Source file exceeds its bound.");
  const raw = fs.readFileSync(target);
  if (createHash("sha256").update(raw).digest("hex") !== location.sha256) throw new Error("Source selection is stale; search again.");
  const { byte_start: start, byte_end: end } = location;
  if (!Number.isSafeInteger(start) || !Number.isSafeInteger(end) || start < 0 || start > end || end > raw.length) {
    throw new Error("Source interval is invalid.");
  }
  const decode = (bytes) => new TextDecoder("utf-8", { fatal: true, ignoreBOM: true }).decode(bytes);
  const point = (bytes) => {
    const lines = decode(bytes).split("\n");
    return { line: lines.length, column: lines.at(-1).length + 1 };
  };
  for (const [key, offset] of [["start", start], ["end", end]]) {
    const expected = point(raw.subarray(0, offset));
    if (location[key]?.line !== expected.line || location[key]?.column !== expected.column) {
      throw new Error("Source editor coordinates differ from the exact bytes.");
    }
  }
  return { path: target, location, text: decode(raw).replace(/\r\n/g, "\n") };
}

async function openSourceLocation(vscode, workspace, location) {
  const target = verifiedSourceTarget(workspace, location);
  const document = await vscode.workspace.openTextDocument(vscode.Uri.file(target.path));
  const editor = await vscode.window.showTextDocument(document, { preview: true });
  const current = verifiedSourceTarget(workspace, location);
  if (document.isDirty || document.getText().replace(/\r\n/g, "\n") !== current.text) {
    throw new Error("The editor buffer differs from the selected source; save and search again.");
  }
  const range = new vscode.Range(location.start.line - 1, location.start.column - 1,
    location.end.line - 1, location.end.column - 1);
  editor.selection = new vscode.Selection(range.start, range.end);
  editor.revealRange(range);
}

module.exports = { invokeSourceAction, verifiedSourceTarget, openSourceLocation };
